### DISCUSSION TITLE
Is match really semantically?

### DISCUSSION THREAD

**Discussion by pjljvandelaar** [ast-grep/ast-grep#1260] [Q&A] [answered]
On https://ast-grep.github.io/advanced/core-concepts.html is written
```
This allows ast-grep to match patterns based on the semantic meaning of the code, not just its surface appearance.
```

Yet, I wonder if it is really the semantic meaning, shouldn't we find two instances when
```
int $VAR = 1;
```
is used to search this code
```Cpp
int i = 1;
int j = 0x1;
```
as both assign the value 1 to the variable, yet use a different representation thereof.

See https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiY3BwIiwicXVlcnkiOiJpbnQgJFZBUiA9IDE7IiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XHJcbiAgICBwYXR0ZXJuOiBpbnQgJEYoJEEsJEIpOyIsInNvdXJjZSI6ImludCBpID0gMTtcbmludCBqID0gMHgxOyJ9

And similarly, 
 shouldn't we find two instances when
```
int $F();
```
is used to search this C++ code
```Cpp
int f();
int g(void);
```
as both declare a function without any parameters.

See https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiY3BwIiwicXVlcnkiOiJpbnQgJEYoKTtcclxuIiwicmV3cml0ZSI6IiIsImNvbmZpZyI6InJ1bGU6XHJcbiAgICBwYXR0ZXJuOiBpbnQgJEYoJEEsJEIpOyIsInNvdXJjZSI6ImludCBmKCk7XG5pbnQgZyh2b2lkKTsifQ==

**Comment by HerringtonDarkholme** [comment]
> the semantic meaning of the code

I will change the wording of the doc. ast-grep is, as its name going, based on syntax tree. It is structural.

If you think ast-grep does not suit your need, you can also try https://semgrep.dev/ or https://about.grit.io/

**Comment by HerringtonDarkholme** [comment, marked as answer]
And also, if you read the whole doc, you should understand ast-grep is base on tree structure. It also has a large Textual vs Structural title on the doc. 

<img width="741" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/12c6d76c-fc92-4376-b319-cfe34eb97484">

Nearly all the paragraphs on the page is discussing the tree-structure. 

Implementation wise, semantics require special logic for each language. The fact that `int f()` and `int f(void)` is equivalent is purely C/C++ semantics. They are **NOT** reflected in the AST.
